src/WebJobs.Extensions.Twilio/Config/TwilioSmsConfiguration.cs

```diff
@@ -0,0 +1,157 @@
+
+        public string From { get; set; }
+
+        public string To { get; set; }
+
+        public void Initialize(ExtensionConfigContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException("context");
+            }
+
+            INameResolver nameResolver = context.Config.GetService<INameResolver>();
+
+            _defaultAccountSid = nameResolver.Resolve(AzureWebJobsTwilioAccountSidKeyName);
+            _defaultAuthToken = nameResolver.Resolve(AzureWebJobsTwilioAccountAuthTokenName);
+
+            IConverterManager converterManager = context.Config.GetService<IConverterManager>();
+            converterManager.AddConverter<JObject, CreateMessageOptions>(CreateMessageOptions);
+
+            BindingFactory factory = new BindingFactory(nameResolver, converterManager);
+            IBindingProvider outputProvider = factory.BindToCollector<TwilioSmsAttribute, CreateMessageOptions>((attr) =>
+            {
+                return new TwilioSmsMessageAsyncCollector(CreateContext(attr));
+            });
+
+            IExtensionRegistry extensions = context.Config.GetService<IExtensionRegistry>();
+            extensions.RegisterBindingRules<TwilioSmsAttribute>(ValidateBinding, nameResolver, outputProvider);
+        }
+
+        private void ValidateBinding(TwilioSmsAttribute attribute, Type type)
+        {
+            string accountSid = Utility.FirstOrDefault(attribute.AccountSidSetting, AccountSid, _defaultAccountSid);
+            string authToken = Utility.FirstOrDefault(attribute.AuthTokenSetting, AuthToken, _defaultAuthToken);
+            if (string.IsNullOrEmpty(accountSid))
+            {
+                ThrowMissingSettingException("AccountSID", AzureWebJobsTwilioAccountSidKeyName, "AccountSID");
+            }
+
+            if (string.IsNullOrEmpty(authToken))
+            {
+                ThrowMissingSettingException("AuthToken", AzureWebJobsTwilioAccountAuthTokenName, "AuthToken");
+            }
+        }
+
+        private TwilioSmsContext CreateContext(TwilioSmsAttribute attribute)
+        {
+            string accountSid = Utility.FirstOrDefault(attribute.AccountSidSetting, AccountSid, _defaultAccountSid);
+            string authToken = Utility.FirstOrDefault(attribute.AuthTokenSetting, AuthToken, _defaultAuthToken);
+
+            TwilioRestClient client = _twilioClientCache.GetOrAdd(new Tuple<string, string>(accountSid, authToken), t => new TwilioRestClient(t.Item1, t.Item2));
+
+            var context = new TwilioSmsContext
+            {
+                Client = client,
+                Body = Utility.FirstOrDefault(attribute.Body, Body),
+                From = Utility.FirstOrDefault(attribute.From, From),
+                To = Utility.FirstOrDefault(attribute.To, To)
+            };
+
+            return context;
+        }
+
+        internal static CreateMessageOptions CreateMessageOptions(JObject messageOptions)
```

Comment:
I'm doing manual deserialization here. I couldn't get Json.NET to deserialize directly into these types for me. I tried lots of different ways, specified the converter array, etc. Open to any suggestions...

Reply:
Does `JsonConvert.PopulateObject` work? Construct CreateMessageOptions via phone number constructor & populate it.

This is the popular answer which looks like manual deserialization to me: https://stackoverflow.com/a/28155770/5915331

Reply:
Yes, that's the first thing I tried. Their custom types for PhoneNumber, enums (that aren't actually enums) were failing to deserialize, even with custom converters added.